File: packages/vm/wasmlib/go/wasmlib/client/arguments.go
package client

import (
	"sort"

	"github.com/iotaledger/wasp/packages/kv/codec"
	"github.com/iotaledger/wasp/packages/vm/wasmlib/go/wasmlib"
)

// The Arguments struct is used to gather all arguments for a smart
// contract function call and encode it into a deterministic byte array
type Arguments struct {
	args map[string][]byte
}

func (a Arguments) set(key string, val []byte) {
	if a.args == nil {
		a.args = make(map[string][]byte)
	}
	a.args[key] = val
}

func (a Arguments) setBase58(key, val string, typeID int32) {
	bytes := Base58Decode(val)
	if len(bytes) != int(wasmlib.TypeSizes[typeID]) {
		panic("invalid byte size")
	}
	a.set(key, bytes)
}

func (a Arguments) Mandatory(key string) {
	if a.args != nil {
		if _, ok := a.args[key]; ok {
			return
		}
	}
	panic("missing mandatory " + key)
}

func (a Arguments) SetAddress(key string, val AgentID) {
	a.setBase58(key, string(val), wasmlib.TYPE_ADDRESS)
}

func (a Arguments) SetAgentID(key string, val AgentID) {
	a.setBase58(key, string(val), wasmlib.TYPE_AGENT_ID)
}

func (a Arguments) SetBool(key string, val bool) {
	bytes := []byte{0}
	if val {
		bytes[0] = 1
	}
	a.set(key, bytes)
}

func (a Arguments) SetBytes(key string, val []byte) {
	a.set(key, val)
}

func (a Arguments) SetColor(key string, val Color) {
	a.setBase58(key, string(val), wasmlib.TYPE_COLOR)
}

func (a Arguments) SetChainID(key string, val ChainID) {
	a.setBase58(key, string(val), wasmlib.TYPE_CHAIN_ID)
}

func (a Arguments) SetHash(key string, val Hash) {
	a.setBase58(key, string(val), wasmlib.TYPE_HASH)
}

func (a Arguments) SetInt8(key string, val int8) {
	a.set(key, []byte{byte(val)})
}

func (a Arguments) SetInt16(key string, val int16) {
	a.setUint64(key, uint64(val), 2)
}

func (a Arguments) SetInt32(key string, val int32) {
	a.setUint64(key, uint64(val), 4)
}

func (a Arguments) SetInt64(key string, val int64) {
	a.setUint64(key, uint64(val), 4)
}

func (a Arguments) SetRequestID(key string, val RequestID) {
	a.setBase58(key, string(val), wasmlib.TYPE_REQUEST_ID)
}

func (a Arguments) SetString(key, val string) {
	a.set(key, []byte(val))
}

func (a Arguments) SetUint8(key string, val uint8) {
	a.set(key, []byte{val})
}

func (a Arguments) SetUint16(key string, val uint16) {
	a.setUint64(key, uint64(val), 2)
}

func (a Arguments) SetUint32(key string, val uint32) {
	a.setUint64(key, uint64(val), 4)
}

func (a Arguments) SetUint64(key string, val uint64) {
	a.setUint64(key, val, 4)
}

func (a Arguments) setUint64(key string, val uint64, size int) {
	bytes := make([]byte, size)
	for i := 0; i < size; i++ {
		bytes[i] = byte(val)
		val >>= 8
	}
	a.set(key, bytes)
}

// Encode returns a byte array that encodes the Arguments as follows:
// Sort all keys in ascending order (very important, because this data
// will be part of the data that will be signed, so the order needs to
// be 100% deterministic). Then emit a 2-byte argument count.
// Next for each argument emit a 2-byte key length, the key prepended
// with a minus sign, a 4-byte value length, and then the value bytes.
func (a Arguments) Encode() []byte {
	keys := make([]string, 0, len(a.args))
	total := 2
	for k, v := range a.args {
		keys = append(keys, k)
		total += 2 + 1 + len(k) + 4 + len(v)
	}
	sort.Strings(keys)

	buf := make([]byte, 0, total)
	buf = append(buf, codec.EncodeUint16(uint16(len(keys)))...)
	for _, k := range keys {
		buf = append(buf, codec.EncodeUint16(uint16(len(k)+1))...)
		buf = append(buf, '-')
		buf = append(buf, []byte(k)...)
		v := a.args[k]
		buf = append(buf, codec.EncodeUint32(uint32(len(v)))...)
		buf = append(buf, v...)
	}
	return buf
}

File: packages/vm/wasmlib/ts/wasmlib/client/arguments.ts
// Copyright 2020 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

import * as wasmlib from "../index"
import * as client from "./index"
import {Base58} from "./crypto";
import {Buffer} from "./buffer";
import {SimpleBufferCursor} from "../../../../../../contracts/wasm/fairroulette/frontend/src/lib/wasp_client";

// The Arguments struct is used to gather all arguments for this smart
// contract function call and encode it into this deterministic byte array
export class Arguments {
	args = new Map<string, client.Bytes>();

	private set(key: string, val: client.Bytes): void {
		this.args.set(key, val);
	}
	
	private setBase58(key: string, val: string, typeID: client.Int32): void {
		let bytes = Base58.decode(val);
		if (bytes.length != wasmlib.TYPE_SIZES[typeID]) {
			client.panic("invalid byte size");
		}
		this.set(key, bytes);
	}

	indexedKey(key: string, index: client.Int32): string {
		return key + "." + index.toString();
	}

	mandatory(key: string): void {
		if (!this.args.has(key)) {
			client.panic("missing mandatory " + key)
		}
	}

	setAddress(key: string, val: client.AgentID): void {
		this.setBase58(key, val, wasmlib.TYPE_ADDRESS);
	}
	
	setAgentID(key: string, val: client.AgentID): void {
		this.setBase58(key, val, wasmlib.TYPE_AGENT_ID);
	}
	
	setBool(key: string, val: boolean): void {
		let bytes = Buffer.alloc(1);
		if (val) {
			bytes.writeUInt8(1, 0);
		}
		this.set(key, bytes)
	}
	
	setBytes(key: string, val: client.Bytes): void {
		this.set(key, Buffer.from(val));
	}
	
	setColor(key: string, val: client.Color): void {
		this.setBase58(key, val, wasmlib.TYPE_COLOR);
	}
	
	setChainID(key: string, val: client.ChainID): void {
		this.setBase58(key, val, wasmlib.TYPE_CHAIN_ID);
	}
	
	setHash(key: string, val: client.Hash): void {
		this.setBase58(key, val, wasmlib.TYPE_HASH);
	}
	
	setInt8(key: string, val: client.Int8): void {
		let bytes = Buffer.alloc(1);
		bytes.writeInt8(val, 0);
		this.set(key, bytes);
	}
	
	setInt16(key: string, val: client.Int16): void {
		let bytes = Buffer.alloc(2);
		bytes.writeInt16LE(val, 0);
		this.set(key, bytes);
	}

	setInt32(key: string, val: client.Int32): void {
		let bytes = Buffer.alloc(4);
		bytes.writeInt32LE(val, 0);
		this.set(key, bytes);
	}
	
	setInt64(key: string, val: client.Int64): void {
		let bytes = Buffer.alloc(8);
		bytes.writeBigInt64LE(val, 0);
		this.set(key, bytes);
	}
	
	setRequestID(key: string, val: client.RequestID): void {
		this.setBase58(key, val, wasmlib.TYPE_REQUEST_ID);
	}
	
	setString(key: string, val: string): void {
		this.set(key, Buffer.from(val));
	}
	
	setUint8(key: string, val: client.Uint8): void {
		let bytes = Buffer.alloc(1);
		bytes.writeUInt8(val, 0);
		this.set(key, bytes);
	}
	
	setUint16(key: string, val: client.Uint16): void {
		let bytes = Buffer.alloc(2);
		bytes.writeUInt16LE(val, 0);
		this.set(key, bytes);
	}
	
	setUint32(key: string, val: client.Uint32): void {
		let bytes = Buffer.alloc(4);
		bytes.writeUInt32LE(val, 0);
		this.set(key, bytes);
	}
	
	setUint64(key: string, val: client.Uint64): void {
		let bytes = Buffer.alloc(8);
		bytes.writeBigUInt64LE(val, 0);
		this.set(key, bytes);
	}

	// Encode returns this byte array that encodes the Arguments as follows:
	// Sort all keys in ascending order (very important, because this data
	// will be part of the data that will be signed, so the order needs to
	// be 100% deterministic). Then emit this 2-byte argument count.
	// Next for each argument emit this 2-byte key length, the key prepended
	// with this minus sign, this 4-byte value length, and then the value bytes.
	encode(): client.Bytes {
		let keys = new Array<string>();
		for (const key of this.args.keys()) {
			keys.push(key);
		}
		keys.sort((lhs, rhs) => lhs.localeCompare(rhs));

		const buf = new SimpleBufferCursor(Buffer.alloc(0));
		buf.writeUInt32LE(keys.length);
		for (const key of keys) {
			let keyBuf = Buffer.from("-" + key);
			buf.writeUInt16LE(keyBuf.length);
			buf.writeBytes(keyBuf);
			let valBuf = this.args.get(key);
			buf.writeUInt32LE(valBuf.length);
			buf.writeBytes(valBuf);
		}
		return buf.buffer;
	}
}
